core/include/drivers/nvmem.h

```diff
@@ -0,0 +1,191 @@
+/**
+ * nvmem_register_provider() - Register a nvmem controller
+ *
+ * @fdt: Device tree to work on
+ * @nodeoffset: Node offset of nvmem cell consumer
+ * @get_dt_nvmem: Callback to match the devicetree nvmem reference with
+ * nvmem_cell
+ * @data: Data which will be passed to the get_dt_nvmem callback
+ * Return a TEE_Result compliant value
+ */
+static inline TEE_Result nvmem_register_provider(const void *fdt,
+						 int nodeoffset,
+						 nvmem_dt_get_func get_dt_nvmem,
+						 void *data)
+{
+	return dt_driver_register_provider(fdt, nodeoffset,
+					   (get_of_device_func)get_dt_nvmem,
+					   data, DT_DRIVER_NVMEM);
+}
+
+/**
+ * nvmem_get_cell_by_name() - Obtain a nvmem cell from its name in the DT node
+ *
+ * @fdt: Device tree to work on
+ * @nodeoffset: Node offset of nvmem cell consumer
+ * @name: name of the nvmem cell defined by property nvmem-cell-names to obtain
+ * from the device tree
+ * @cell: Pointer filled with the retrieved cell, must be freed after use
+   using nvmem_put_cell()
+ * Return a TEE_Result compliant value
+ */
+TEE_Result nvmem_get_cell_by_name(const void *fdt, int nodeoffset,
+				  const char *name, struct nvmem_cell **cell);
+
+/**
+ * nvmem_get_cell_by_index() - Obtain a nvmem cell from property nvmem-cells
+ *
+ * @fdt: Device tree to work on
+ * @nodeoffset: Node offset of nvmem cell consumer
+ * @index: Index of the nvmem cell to obtain from device-tree
+ * @out_cell: Pointer filled with the retrieved cell, must be freed after use
+ * using nvmem_put_cell()
+ * Return a TEE_Result compliant value
+ */
+static inline TEE_Result nvmem_get_cell_by_index(const void *fdt,
+						 int nodeoffset,
+						 unsigned int index,
+						 struct nvmem_cell **out_cell)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	void *cell = NULL;
+
+	res = dt_driver_device_from_node_idx_prop("nvmem-cells", fdt,
+						  nodeoffset, index,
+						  DT_DRIVER_NVMEM, &cell);
+	if (!res)
+		*out_cell = cell;
+
+	return res;
+}
+
+/**
+ * nvmem_cell_parse_dt() - Parse device-tree information to fill a nvmem cell
+ *
+ * @fdt: Device tree to work on
+ * @nodeoffset: Node offset of the nvmem cell controller
+ * @cell: Pointer to cell that will be filled
+ */
+TEE_Result nvmem_cell_parse_dt(const void *fdt, int nodeoffset,
+			       struct nvmem_cell *cell);
+
+/**
+ * nvmem_put_cell() - Free resource allocated from nvmem_get_cell_by_*()
+ *
+ * @cell: Cell to be freed
+ */
+static inline void nvmem_put_cell(struct nvmem_cell *cell)
+{
+	if (cell->ops->free_cell)
+		cell->ops->free_cell(cell);
+}
+
+/*
+ * nvmem_cell_alloc_read() - Allocate and read data from a nvmem cell
+ * @cell: Cell to read from nvmem
+ * @data: Output allocated buffer where nvmem cell data are stored upon success,
+ * shall be released with free().
+ * @len: Output byte size of the nvmem cell data read
+ */
+static inline TEE_Result nvmem_cell_alloc_read(struct nvmem_cell *cell,
+					       uint8_t **data, size_t *len)
+{
+	if (!cell->ops->alloc_read_cell)
```

Comment:
I propose to slightly change this API:
* Let's name it `nvmem_cell_malloc_and_read()` so that it's clear the buffer is allocated with `malloc()` and shall be freed with `free()`.
* There is no need for each driver to implement the allocation sequence, so let's change the ops handler from `.alloc_read_cell` to `.read_cell` this function and add a simpler `nvmem_cell_read()` API function:
```c
struct nvmem_ops {
	/*
	 * Read data from an NVMEM cell.
	 * @cell: Cell to read data from
	 * @data: Output buffer of size greater or equal to @cell->size
	 */
	TEE_Result (*read_cell)(struct nvmem_cell *cell, uint8_t *data);

	(...)
};

(...)

/*
 * nvmem_cell_read() - Read data from a nvmem cell
 * @cell: Cell to read from nvmem
 * @data: Output data read from the cell upon success, byte size >= @cell->size 
 */
static inline TEE_Result nvmem_cell_read(struct nvmem_cell *cell,
					 uint8_t *data)
{
	if (!cell->ops->read_cell)
		return TEE_ERROR_NOT_SUPPORTED;

	return cell->ops->read_cell(cell, data);
}

/*
 * nvmem_cell_malloc_and_read() - Allocate and read data from a nvmem cell
 * @cell: Cell to read from nvmem
 * @data: Output allocated buffer where nvmem cell data are stored upon success
 *
 * Upon success, the output buffer is allocated with malloc(). Caller is
 * responsible for freeing the buffer with free() if needed.
 */
TEE_Result nvmem_cell_malloc_and_read(struct nvmem_cell *cell,
				      uint8_t **out_data);
```